prereise/gather/flexibilitydata/geo/geo_data.py

```diff
@@ -0,0 +1,415 @@
+    param string cache_path: folder to store processed cache files
+    :return: (*None*)
+    """
+    
+    df_raw = pd.read_csv(raw_data_path+"/county_to_zip.csv")
+    all_fips = df_raw['county'].astype('int32')
+    all_zip = df_raw['zip'].astype('int32')
+    all_weights = df_raw['tot_ratio']
+
+    # create zip -> counties mapping
+    zip2fips = {}
+    
+    for i in pd.unique(all_zip):
+        idx = all_zip.index[all_zip == i]
+        
+        cty = all_fips[idx].tolist()
+        wgt = all_weights[idx].tolist()
+        
+        zip2fips.update({i: (cty, wgt)})
+
+    with open(cache_path+"/zip2fips.pkl", 'wb') as fh:
+        pkl.dump(zip2fips, fh)
+
+    # create county -> zips mapping
+    fip2zips = {}
+    
+    for i in pd.unique(all_fips):
+        idx = all_fips.index[all_fips == i]
+        
+        zips = all_zip[idx].tolist()
+        wgt = all_weights[idx].tolist()
+        
+        fip2zips.update({i: (zips, wgt)})
+
+
+    with open(cache_path+"/fip2zips.pkl", 'wb') as fh:
+        pkl.dump(fip2zips, fh)
+
+
+def get_fips_population(raw_data_path, cache_path):
+    """Match county population and county FIPS data to produce concise FIPS population
+        save to a dictonary in cache folder with key being 5-digit FIPS codes.
+
+    param string raw_data_path: folder that contains downloaded raw data
+    param string cache_path: folder to store processed cache files
+    :return: (*None*)
+    """
+
+    cty_pop_df = pd.read_csv(raw_data_path+"/county_population.csv", encoding='cp1252')
+    cty_name_df = pd.read_csv(raw_data_path+"/county_fips_master.csv", encoding='cp1252')
+
+    pops = np.zeros(len(cty_name_df.index))
+    for cty in cty_name_df.index:
+        name = cty_name_df['county_name'][cty]
+        state = cty_name_df['state_name'][cty]
+
+        tmp = cty_pop_df.loc[(cty_pop_df["CTYNAME"]==name) & (cty_pop_df["STNAME"]==state), "POPESTIMATE2019"]
+        # if no data
+        if len(tmp) == 0:
+            pops[cty] = 0
+        elif len(tmp) == 1:
+            pops[cty] = tmp
+        else:
+            pops[cty] = tmp.to_list()[0]
+            
+    cty_name_df["population"] = pops
+
+    new_df = cty_name_df.loc[:, ["fips", "county_name", "population"]]
+    new_df.to_csv(cache_path+"/fips_population.csv", index=False)
+
+
+
+def get_zip_population(raw_data_path, cache_path):
+    """Compute population of each ZIP code using percentage share and FIPS population
+        save to a dictonary in cache folder with key being zip codes.
+        
+    param string raw_data_path: folder that contains downloaded raw data
+    param string cache_path: folder to store processed cache files
+    :return: (*None*)
+    """
+    
+    fips_pop_df = pd.read_csv(cache_path+"/fips_population.csv")
+    
+    with open(cache_path+"/zip2fips.pkl", 'rb') as fh:
+        zip2fips = pkl.load(fh)
+
+    zips = list(zip2fips.keys())
+    zips.sort()
```

Comment:
Can these lines be combined into one?

Reply:
Yes

Reply:
actually not here.. `zips = list(zip2fips.keys()).sort()` sets `zips` to the return value of `list.sort()` and is not the sorted list itself. The Python list functions behave differently as their Pandas counterpart (in which most functions return a copy of the list after operation) so they have to be separated.

Reply:
Would `zips = sorted(zip2fips.keys())` do what we want?

Reply:
great, I didn't know this exist.